executor/foreign_key.go

```diff
@@ -0,0 +1,488 @@
+	err = doLockKeys(ctx, fkc.ctx, lockCtx, fkc.toBeLockedKeys...)
+	// doLockKeys may set TxnCtx.ForUpdate to 1, then if the lock meet write conflict, TiDB can't retry for update.
+	// So reset TxnCtx.ForUpdate to 0 then can be retry if meet write conflict.
+	atomic.StoreUint32(&sessVars.TxnCtx.ForUpdate, forUpdate)
+	return err
+}
+
+func (fkc *FKCheckExec) buildCheckKeyFromFKValue(sc *stmtctx.StatementContext, vals []types.Datum) (key kv.Key, isPrefix bool, err error) {
+	if fkc.IdxIsPrimaryKey {
+		handleKey, err := fkc.buildHandleFromFKValues(sc, vals)
+		if err != nil {
+			return nil, false, err
+		}
+		key := tablecodec.EncodeRecordKey(fkc.Tbl.RecordPrefix(), handleKey)
+		if fkc.IdxIsExclusive {
+			return key, false, nil
+		}
+		return key, true, nil
+	}
+	key, distinct, err := fkc.Idx.GenIndexKey(sc, vals, nil, nil)
+	if err != nil {
+		return nil, false, err
+	}
+	if distinct && fkc.IdxIsExclusive {
+		return key, false, nil
+	}
+	return key, true, nil
+}
+
+func (fkc *FKCheckExec) buildHandleFromFKValues(sc *stmtctx.StatementContext, vals []types.Datum) (kv.Handle, error) {
+	if len(vals) == 1 && fkc.Idx == nil {
+		return kv.IntHandle(vals[0].GetInt64()), nil
+	}
+	handleBytes, err := codec.EncodeKey(sc, nil, vals...)
+	if err != nil {
+		return nil, err
+	}
+	return kv.NewCommonHandle(handleBytes)
+}
+
+func (fkc *FKCheckExec) checkKeys(ctx context.Context, txn kv.Transaction) error {
+	if len(fkc.toBeCheckedKeys) == 0 {
+		return nil
+	}
+	err := fkc.prefetchKeys(ctx, txn, fkc.toBeCheckedKeys)
+	if err != nil {
+		return err
+	}
+	for _, k := range fkc.toBeCheckedKeys {
+		err = fkc.checkKey(ctx, txn, k)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (fkc *FKCheckExec) prefetchKeys(ctx context.Context, txn kv.Transaction, keys []kv.Key) error {
+	// Fill cache using BatchGet
+	_, err := txn.BatchGet(ctx, keys)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (fkc *FKCheckExec) checkKey(ctx context.Context, txn kv.Transaction, k kv.Key) error {
+	if fkc.CheckExist {
+		return fkc.checkKeyExist(ctx, txn, k)
+	}
+	return fkc.checkKeyNotExist(ctx, txn, k)
+}
+
+func (fkc *FKCheckExec) checkKeyExist(ctx context.Context, txn kv.Transaction, k kv.Key) error {
+	_, err := txn.Get(ctx, k)
+	if err == nil {
+		fkc.toBeLockedKeys = append(fkc.toBeLockedKeys, k)
+		return nil
+	}
+	if kv.IsErrNotFound(err) {
+		return fkc.FailedErr
+	}
+	return err
+}
+
+func (fkc *FKCheckExec) checkKeyNotExist(ctx context.Context, txn kv.Transaction, k kv.Key) error {
+	_, err := txn.Get(ctx, k)
+	if err == nil {
+		return fkc.FailedErr
+	}
+	if kv.IsErrNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+func (fkc *FKCheckExec) checkIndexKeys(ctx context.Context, txn kv.Transaction) error {
+	if len(fkc.toBeCheckedPrefixKeys) == 0 {
+		return nil
+	}
+	memBuffer := txn.GetMemBuffer()
+	snap := txn.GetSnapshot()
+	snap.SetOption(kv.ScanBatchSize, 2)
+	defer func() {
+		snap.SetOption(kv.ScanBatchSize, 256)
```

Comment:
I see the `const defaultScanBatchSize = 128`

Reply:
after https://github.com/tikv/client-go/pull/588 merged, I will use the `DefaultScanBatchSize` instead `256` here.

Reply:
Can we use `DefaultScanBatchSize` instead of a magic number?

Reply:
done.